### PR TITLE
fix(waiter): use a buffered channel to fix race condition

### PR DESCRIPTION
**Pull request by ctlong**
# Description

We discovered a race condition where `Waiter.Set` and `Waiter.Next` could both be called once, but `Waiter.Next` would never resolve with the result.

The reason for this is that `sync.Cond` does not appear to retain any kind of state, so if `w.c.Broadcast` is called first, `w.c.Wait` does not know that it can continue.

By using a buffered channel instead we can get around that issue.

According to the benchmark tests this is actually faster too:
```
$ go test -bench=Waiter -run=NoTest
goos: darwin
goarch: arm64
pkg: code.cloudfoundry.org/go-diodes
BenchmarkOneToOneWaiter-10         	14808483	        73.74 ns/op
BenchmarkManyToOneWaiter-10        	17037690	        70.63 ns/op
BenchmarkOneToOneWaiterDrain-10    	18203583	        55.37 ns/op
PASS
ok  	code.cloudfoundry.org/go-diodes	3.747s
$ git co fix/waiter/race-condition
$ go test -bench=Waiter -run=NoTest
goos: darwin
goarch: arm64
pkg: code.cloudfoundry.org/go-diodes
BenchmarkOneToOneWaiter-10         	31560381	        38.86 ns/op
BenchmarkManyToOneWaiter-10        	22670631	        49.88 ns/op
BenchmarkOneToOneWaiterDrain-10    	23677998	        48.81 ns/op
PASS
ok  	code.cloudfoundry.org/go-diodes	5.239s
```
```bash
$ go test -bench=Waiter -run=NoTest
goos: linux
goarch: amd64
pkg: code.cloudfoundry.org/go-diodes
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkOneToOneWaiter-8                5920287               195.4 ns/op
BenchmarkManyToOneWaiter-8               5823700               192.6 ns/op
BenchmarkOneToOneWaiterDrain-8           7738105               147.2 ns/op
PASS
ok      code.cloudfoundry.org/go-diodes 4.261s
$ git co fix/waiter/race-condition
$ go test -bench=Waiter -run=NoTest
goos: linux
goarch: amd64
pkg: code.cloudfoundry.org/go-diodes
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkOneToOneWaiter-8                7039336               148.9 ns/op
BenchmarkManyToOneWaiter-8               7326846               164.4 ns/op
BenchmarkOneToOneWaiterDrain-8           8490176               127.3 ns/op
PASS
ok      code.cloudfoundry.org/go-diodes 3.956s
```
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing performed?

- [x] Unit tests
- [ ] Integration tests
- [ ] Acceptance tests

## Checklist:

- [x] This PR is being made against the `main` branch, or relevant version branch
- [ ] I have made corresponding changes to the documentation
- [ ] I have added testing for my changes

### CODE REVIEW

**Comment by ctlong** [comment]
We discovered the race condition while investigating failing loggregator unit tests: https://github.com/cloudfoundry/loggregator-release/actions/runs/6226410984.

File: waiter.go [hunk 1]
```diff
@@ -30,54 +28,41 @@ func WithWaiterContext(ctx context.Context) WaiterConfigOption {
...
 
-	go func() {
-		<-w.ctx.Done()
-		w.c.Broadcast()
-	}()
-
 	return w
 }
 
 // Set invokes the wrapped diode's Set with the given data and uses Broadcast
```
Comment (ctlong) [inline waiter.go:41]: Update comment

File: waiter.go [hunk 2]
```diff
@@ -2,15 +2,13 @@ package diodes
 
 import (
 	"context"
-	"sync"
 )
 
 // Waiter will use a conditional mutex to alert the reader to when data is
```
Comment (ctlong) [inline waiter.go:7]: Update comment

File: waiter.go [hunk 3]
```diff
@@ -30,54 +28,41 @@ func WithWaiterContext(ctx context.Context) WaiterConfigOption {
...
-	w.c.Broadcast()
+	select {
+	case w.c <- struct{}{}:
+	default:
+		return
+	}
 }
 
 // Next returns the next data point on the wrapped diode. If there is not any
 // new data, it will Wait for set to be called or the context to be done.
```
Comment (ctlong) [inline waiter.go:53]: Update comment

File: waiter.go [hunk 4]
```diff
@@ -30,54 +28,41 @@ func WithWaiterContext(ctx context.Context) WaiterConfigOption {
...
-
 	for {
 		data, ok := w.Diode.TryNext()
 		if !ok {
-			if w.isDone() {
+			select {
+			case <-w.ctx.Done():
 				return nil
+			case <-w.c:
+				continue
```
Comment (ctlong) [inline waiter.go:63]: Extract function?

File: waiter.go [hunk 5]
```diff
@@ -30,54 +28,41 @@ func WithWaiterContext(ctx context.Context) WaiterConfigOption {
...
 // Set invokes the wrapped diode's Set with the given data and uses Broadcast
 // to wake up any readers.
 func (w *Waiter) Set(data GenericDataType) {
 	w.Diode.Set(data)
-	w.c.Broadcast()
+	select {
+	case w.c <- struct{}{}:
+	default:
+		return
+	}
```
Comment (ctlong) [inline waiter.go:49]: Extract function?

**Review by chombium — APPROVED** [review]
LGTM!
@ctlong Thanks for taking care of the suggested changes this fast

**Comment by ctlong** [comment]
@mkocher and I wanted to see how this affected a running CF-D, so we took some CPU samples off of a Diego Cell VM that had 5 loggregator-load-generator-go apps running with a speed of 10 microseconds between log lines:
```
$ while true; do top -b -n1 -o '%CPU' | grep -E 'forwarder-agent|syslog-agent|loggregator|metrics' | awk '{print $NF " " $9}' | xargs -n2 printf "%s\t%s\t%s\n" "$(date +%H:%M:%S)"; sleep 0.1; done
23:53:12        loggregator-age 23.5
23:53:12        syslog-agent    72.2
23:53:12        forwarder-agent 55.6
23:53:12        metrics-agent   33.3
23:53:12        loggregator-age 16.7
23:53:12        forwarder-agent 58.8
23:53:12        syslog-agent    58.8
23:53:12        loggregator-age 47.1
23:53:12        metrics-agent   29.4
23:53:13        forwarder-agent 64.7
23:53:13        syslog-agent    52.9
23:53:13        loggregator-age 41.2
23:53:13        metrics-agent   23.5
23:53:13        forwarder-agent 64.7
23:53:13        syslog-agent    58.8
23:53:13        loggregator-age 35.3
23:53:13        metrics-agent   23.5
23:53:13        forwarder-agent 64.7
23:53:13        syslog-agent    52.9
23:53:13        loggregator-age 29.4
23:53:13        metrics-agent   17.6
23:53:14        forwarder-agent 56.2
23:53:14        syslog-agent    50.0
23:53:14        loggregator-age 18.8
23:53:14        metrics-agent   18.8
23:53:14        forwarder-agent 58.8
23:53:14        syslog-agent    52.9
23:53:14        loggregator-age 23.5
23:53:14        metrics-agent   23.5
23:53:14        forwarder-agent 70.6
23:53:14        loggregator-age 52.9
23:53:14        syslog-agent    52.9
23:53:14        metrics-agent   23.5
23:53:14        forwarder-agent 58.8
23:53:14        syslog-agent    58.8
23:53:14        loggregator-age 41.2
23:53:14        metrics-agent   29.4
23:53:15        syslog-agent    68.8
23:53:15        forwarder-agent 62.5
23:53:15        metrics-agent   37.5
23:53:15        loggregator-age 12.5
23:53:15        forwarder-agent 81.2
23:53:15        syslog-agent    75.0
23:53:15        loggregator-age 43.8
23:53:15        metrics-agent   25.0
23:53:15        forwarder-agent 58.8
23:53:15        syslog-agent    52.9
23:53:15        metrics-agent   29.4
23:53:15        loggregator-age 5.9
23:53:16        forwarder-agent 75.0
23:53:16        syslog-agent    62.5
23:53:16        loggregator-age 50.0
23:53:16        metrics-agent   31.2
```
Then, after redeploying a dev version of loggregator-agent with this change to the diodes library:
```
$ while true; do top -b -n1 -o '%CPU' | grep -E 'forwarder-agent|syslog-agent|metrics' | awk '{print $NF " " $9}' | xargs -n2 printf "%s\t%s\t%s\n" "$(date +%H:%M:%S)"; sleep 0.1; done
22:23:55        forwarder-agent 64.7                                                                                                                                                                                                                                  [0/3230]
22:23:55        syslog-agent    52.9
22:23:55        metrics-agent   23.5
22:23:56        syslog-agent    62.5
22:23:56        forwarder-agent 56.2
22:23:56        metrics-agent   25.0
22:23:56        forwarder-agent 58.8
22:23:56        syslog-agent    47.1
22:23:56        metrics-agent   23.5
22:23:56        forwarder-agent 64.7
22:23:56        syslog-agent    64.7
22:23:56        metrics-agent   29.4
22:23:57        forwarder-agent 50.0
22:23:57        syslog-agent    50.0
22:23:57        metrics-agent   18.8
22:23:57        forwarder-agent 64.7
22:23:57        syslog-agent    64.7
22:23:57        metrics-agent   23.5
22:23:57        forwarder-agent 64.7
22:23:57        syslog-agent    52.9
22:23:57        metrics-agent   23.5
22:23:57        forwarder-agent 58.8
22:23:57        syslog-agent    47.1
22:23:57        metrics-agent   23.5
22:23:58        forwarder-agent 76.5
22:23:58        syslog-agent    58.8
22:23:58        metrics-agent   29.4
22:23:58        forwarder-agent 52.9
22:23:58        syslog-agent    41.2
22:23:58        metrics-agent   23.5
22:23:58        forwarder-agent 58.8
22:23:58        syslog-agent    52.9
22:23:58        metrics-agent   17.6
22:23:59        forwarder-agent 64.7
22:23:59        syslog-agent    64.7
22:23:59        metrics-agent   29.4
22:23:59        forwarder-agent 66.7
22:23:59        syslog-agent    61.1
22:23:59        metrics-agent   22.2
22:23:59        forwarder-agent 70.6
22:23:59        syslog-agent    47.1
22:23:59        metrics-agent   23.5
22:24:00        forwarder-agent 70.6
22:24:00        syslog-agent    64.7
22:24:00        metrics-agent   29.4
22:24:00        forwarder-agent 58.8
22:24:00        syslog-agent    47.1
22:24:00        metrics-agent   23.5
22:24:00        forwarder-agent 64.7
22:24:00        syslog-agent    52.9
22:24:00        metrics-agent   17.6
22:24:01        syslog-agent    64.7
22:24:01        forwarder-agent 52.9
```

We saw that every agent decreased its CPU usage by several percentage points

**Review by mkocher — APPROVED** [review]

**Pull request merged**